File: sdks/go/crypto.go
package acp

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/ed25519"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"sort"

	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"
)

func toBase64URL(bytes []byte) string {
	return base64.RawURLEncoding.EncodeToString(bytes)
}

func fromBase64URL(value string) ([]byte, error) {
	decoded, err := base64.RawURLEncoding.DecodeString(value)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("invalid base64 value: %v", err))
	}
	return decoded, nil
}

func asFixedBytes(value []byte, length int, label string) ([]byte, error) {
	if len(value) != length {
		return nil, CryptoError(fmt.Sprintf("invalid %s length", label))
	}
	return value, nil
}

func SHA256Hex(input []byte) string {
	sum := sha256.Sum256(input)
	return hex.EncodeToString(sum[:])
}

func GenerateEd25519Keypair() (privateKey string, publicKey string, err error) {
	seed := make([]byte, ed25519.SeedSize)
	if _, err := rand.Read(seed); err != nil {
		return "", "", CryptoError(fmt.Sprintf("unable to generate Ed25519 seed: %v", err))
	}
	key := ed25519.NewKeyFromSeed(seed)
	public := key.Public().(ed25519.PublicKey)
	return toBase64URL(seed), toBase64URL(public), nil
}

func Ed25519PublicFromPrivate(signingPrivateKeyB64 string) (string, error) {
	seed, err := fromBase64URL(signingPrivateKeyB64)
	if err != nil {
		return "", err
	}
	seed, err = asFixedBytes(seed, ed25519.SeedSize, "Ed25519 private key")
	if err != nil {
		return "", err
	}
	key := ed25519.NewKeyFromSeed(seed)
	return toBase64URL(key.Public().(ed25519.PublicKey)), nil
}

func GenerateX25519Keypair() (privateKey string, publicKey string, err error) {
	privateBytes := make([]byte, 32)
	if _, err := rand.Read(privateBytes); err != nil {
		return "", "", CryptoError(fmt.Sprintf("unable to generate X25519 private key: %v", err))
	}
	publicBytes, err := curve25519.X25519(privateBytes, curve25519.Basepoint)
	if err != nil {
		return "", "", CryptoError(fmt.Sprintf("unable to derive X25519 public key: %v", err))
	}
	return toBase64URL(privateBytes), toBase64URL(publicBytes), nil
}

func X25519PublicFromPrivate(encryptionPrivateKeyB64 string) (string, error) {
	privateBytes, err := fromBase64URL(encryptionPrivateKeyB64)
	if err != nil {
		return "", err
	}
	privateBytes, err = asFixedBytes(privateBytes, 32, "X25519 private key")
	if err != nil {
		return "", err
	}
	publicBytes, err := curve25519.X25519(privateBytes, curve25519.Basepoint)
	if err != nil {
		return "", CryptoError(fmt.Sprintf("unable to derive X25519 public key: %v", err))
	}
	return toBase64URL(publicBytes), nil
}

func SignBytes(data []byte, signingPrivateKeyB64 string) (string, error) {
	seed, err := fromBase64URL(signingPrivateKeyB64)
	if err != nil {
		return "", err
	}
	seed, err = asFixedBytes(seed, ed25519.SeedSize, "Ed25519 private key")
	if err != nil {
		return "", err
	}
	signingKey := ed25519.NewKeyFromSeed(seed)
	signature := ed25519.Sign(signingKey, data)
	return toBase64URL(signature), nil
}

func VerifySignature(data []byte, signatureB64 string, signingPublicKeyB64 string) bool {
	signature, err := fromBase64URL(signatureB64)
	if err != nil || len(signature) != ed25519.SignatureSize {
		return false
	}
	publicKey, err := fromBase64URL(signingPublicKeyB64)
	if err != nil || len(publicKey) != ed25519.PublicKeySize {
		return false
	}
	return ed25519.Verify(ed25519.PublicKey(publicKey), data, signature)
}

func EnvelopeAAD(envelope Envelope) ([]byte, error) {
	return CanonicalJSONBytes(map[string]any{
		"acp_version":  envelope.ACPVersion,
		"message_id":   envelope.MessageID,
		"operation_id": envelope.OperationID,
		"sender":       envelope.Sender,
		"recipients":   envelope.Recipients,
	})
}

func deriveWrapKey(sharedSecret []byte, recipient string) ([]byte, error) {
	reader := hkdf.New(sha256.New, sharedSecret, []byte{}, []byte("acp-v1-wrap:"+recipient))
	key := make([]byte, 32)
	if _, err := io.ReadFull(reader, key); err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to derive wrap key: %v", err))
	}
	return key, nil
}

func encryptAESGCM(key []byte, nonce []byte, plaintext []byte, aad []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to create AES cipher: %v", err))
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to create AES-GCM cipher: %v", err))
	}
	if len(nonce) != gcm.NonceSize() {
		return nil, CryptoError("invalid nonce length")
	}
	return gcm.Seal(nil, nonce, plaintext, aad), nil
}

func decryptAESGCM(key []byte, nonce []byte, ciphertext []byte, aad []byte) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to create AES cipher: %v", err))
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to create AES-GCM cipher: %v", err))
	}
	if len(nonce) != gcm.NonceSize() {
		return nil, CryptoError("invalid nonce length")
	}
	plaintext, err := gcm.Open(nil, nonce, ciphertext, aad)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to decrypt AES-GCM payload: %v", err))
	}
	return plaintext, nil
}

func randomBytes(length int) ([]byte, error) {
	out := make([]byte, length)
	if _, err := rand.Read(out); err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to generate secure random bytes: %v", err))
	}
	return out, nil
}

func EncryptForRecipients(
	payload map[string]any,
	envelope Envelope,
	recipientEncryptionPublicKeys map[string]string,
) (ProtectedPayload, error) {
	plaintext, err := CanonicalJSONBytes(payload)
	if err != nil {
		return ProtectedPayload{}, err
	}
	contentKey, err := randomBytes(32)
	if err != nil {
		return ProtectedPayload{}, err
	}
	payloadNonce, err := randomBytes(12)
	if err != nil {
		return ProtectedPayload{}, err
	}
	aad, err := EnvelopeAAD(envelope)
	if err != nil {
		return ProtectedPayload{}, err
	}
	payloadCiphertext, err := encryptAESGCM(contentKey, payloadNonce, plaintext, aad)
	if err != nil {
		return ProtectedPayload{}, err
	}
	ephemeralPrivate, ephemeralPublic, err := GenerateX25519Keypair()
	if err != nil {
		return ProtectedPayload{}, err
	}
	ephemeralPrivateBytes, _ := fromBase64URL(ephemeralPrivate)
	wrapped := make([]WrappedContentKey, 0, len(recipientEncryptionPublicKeys))
	recipients := make([]string, 0, len(recipientEncryptionPublicKeys))
	for recipient := range recipientEncryptionPublicKeys {
		recipients = append(recipients, recipient)
	}
	sort.Strings(recipients)
	for _, recipient := range recipients {
		recipientPublicB64 := recipientEncryptionPublicKeys[recipient]
		recipientPublicBytes, err := fromBase64URL(recipientPublicB64)
		if err != nil {
			return ProtectedPayload{}, err
		}
		recipientPublicBytes, err = asFixedBytes(recipientPublicBytes, 32, "X25519 public key")
		if err != nil {
			return ProtectedPayload{}, err
		}
		sharedSecret, err := curve25519.X25519(ephemeralPrivateBytes, recipientPublicBytes)
		if err != nil {
			return ProtectedPayload{}, CryptoError(fmt.Sprintf("unable to derive X25519 shared secret: %v", err))
		}
		wrapKey, err := deriveWrapKey(sharedSecret, recipient)
		if err != nil {
			return ProtectedPayload{}, err
		}
		wrapNonce, err := randomBytes(12)
		if err != nil {
			return ProtectedPayload{}, err
		}
		wrappedCiphertext, err := encryptAESGCM(wrapKey, wrapNonce, contentKey, []byte(envelope.MessageID))
		if err != nil {
			return ProtectedPayload{}, err
		}
		wrapped = append(wrapped, WrappedContentKey{
			Recipient:          recipient,
			EphemeralPublicKey: ephemeralPublic,
			Nonce:              toBase64URL(wrapNonce),
			Ciphertext:         toBase64URL(wrappedCiphertext),
		})
	}
	return ProtectedPayload{
		Nonce:              toBase64URL(payloadNonce),
		Ciphertext:         toBase64URL(payloadCiphertext),
		WrappedContentKeys: wrapped,
		PayloadHash:        SHA256Hex(payloadCiphertext),
		SignatureKID:       "",
		Signature:          "",
	}, nil
}

func messageSignatureInput(envelope Envelope, protectedPayload ProtectedPayload) ([]byte, error) {
	wrapped := append([]WrappedContentKey{}, protectedPayload.WrappedContentKeys...)
	sort.Slice(wrapped, func(left int, right int) bool {
		return wrapped[left].Recipient < wrapped[right].Recipient
	})
	signable := map[string]any{
		"envelope": envelope,
		"protected": map[string]any{
			"nonce":                protectedPayload.Nonce,
			"ciphertext":           protectedPayload.Ciphertext,
			"wrapped_content_keys": wrapped,
			"payload_hash":         protectedPayload.PayloadHash,
			"signature_kid":        protectedPayload.SignatureKID,
		},
	}
	return CanonicalJSONBytes(signable)
}

func SignProtectedPayload(
	envelope Envelope,
	protectedPayload ProtectedPayload,
	signingPrivateKeyB64 string,
	signatureKID string,
) (ProtectedPayload, error) {
	signed := protectedPayload
	signed.SignatureKID = signatureKID
	input, err := messageSignatureInput(envelope, signed)
	if err != nil {
		return ProtectedPayload{}, err
	}
	signature, err := SignBytes(input, signingPrivateKeyB64)
	if err != nil {
		return ProtectedPayload{}, err
	}
	signed.Signature = signature
	return signed, nil
}

func VerifyProtectedPayloadSignature(
	envelope Envelope,
	protectedPayload ProtectedPayload,
	senderSigningPublicKeyB64 string,
) bool {
	if protectedPayload.Signature == "" {
		return false
	}
	input, err := messageSignatureInput(envelope, protectedPayload)
	if err != nil {
		return false
	}
	return VerifySignature(input, protectedPayload.Signature, senderSigningPublicKeyB64)
}

func DecryptForRecipient(
	envelope Envelope,
	protectedPayload ProtectedPayload,
	recipientID string,
	recipientEncryptionPrivateKeyB64 string,
) (map[string]any, error) {
	var wrapped *WrappedContentKey
	for index := range protectedPayload.WrappedContentKeys {
		if protectedPayload.WrappedContentKeys[index].Recipient == recipientID {
			wrapped = &protectedPayload.WrappedContentKeys[index]
			break
		}
	}
	if wrapped == nil {
		return nil, CryptoError(fmt.Sprintf("No wrapped content key available for recipient %s", recipientID))
	}
	recipientPrivate, err := fromBase64URL(recipientEncryptionPrivateKeyB64)
	if err != nil {
		return nil, err
	}
	recipientPrivate, err = asFixedBytes(recipientPrivate, 32, "X25519 private key")
	if err != nil {
		return nil, err
	}
	ephemeralPublic, err := fromBase64URL(wrapped.EphemeralPublicKey)
	if err != nil {
		return nil, err
	}
	ephemeralPublic, err = asFixedBytes(ephemeralPublic, 32, "X25519 public key")
	if err != nil {
		return nil, err
	}
	sharedSecret, err := curve25519.X25519(recipientPrivate, ephemeralPublic)
	if err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to derive X25519 shared secret: %v", err))
	}
	wrapKey, err := deriveWrapKey(sharedSecret, recipientID)
	if err != nil {
		return nil, err
	}
	wrapNonce, err := fromBase64URL(wrapped.Nonce)
	if err != nil {
		return nil, err
	}
	wrapNonce, err = asFixedBytes(wrapNonce, 12, "wrapped nonce")
	if err != nil {
		return nil, err
	}
	wrappedCiphertext, err := fromBase64URL(wrapped.Ciphertext)
	if err != nil {
		return nil, err
	}
	contentKey, err := decryptAESGCM(wrapKey, wrapNonce, wrappedCiphertext, []byte(envelope.MessageID))
	if err != nil {
		return nil, err
	}
	payloadNonce, err := fromBase64URL(protectedPayload.Nonce)
	if err != nil {
		return nil, err
	}
	payloadNonce, err = asFixedBytes(payloadNonce, 12, "payload nonce")
	if err != nil {
		return nil, err
	}
	payloadCiphertext, err := fromBase64URL(protectedPayload.Ciphertext)
	if err != nil {
		return nil, err
	}
	aad, err := EnvelopeAAD(envelope)
	if err != nil {
		return nil, err
	}
	plaintext, err := decryptAESGCM(contentKey, payloadNonce, payloadCiphertext, aad)
	if err != nil {
		return nil, err
	}
	var parsed map[string]any
	if err := json.Unmarshal(plaintext, &parsed); err != nil {
		return nil, CryptoError(fmt.Sprintf("unable to parse decrypted payload: %v", err))
	}
	return parsed, nil
}

File: sdks/typescript/src/crypto.ts
/*
 * Copyright 2026 ACP Project
 * Licensed under the Apache License, Version 2.0
 * See LICENSE file for details.
 */

import {
  createCipheriv,
  createDecipheriv,
  createHash,
  hkdfSync,
  randomBytes
} from "node:crypto";
import nacl from "tweetnacl";
import { cryptoError } from "./errors";
import { canonicalJsonBytes, JsonMap, JsonValue, parseJsonMap } from "./jsonSupport";
import { Envelope, ProtectedPayload, WrappedContentKey } from "./messages";

function toBase64Url(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("base64url");
}

function fromBase64Url(value: string): Uint8Array {
  try {
    return Buffer.from(value, "base64url");
  } catch (error) {
    throw cryptoError(`invalid base64 value: ${String(error)}`);
  }
}

function asFixedBytes(value: Uint8Array, length: number, label: string): Uint8Array {
  if (value.length !== length) {
    throw cryptoError(`invalid ${label} length`);
  }
  return value;
}

export function sha256Hex(input: Uint8Array): string {
  return createHash("sha256").update(input).digest("hex");
}

export function generateEd25519Keypair(): { private_key: string; public_key: string } {
  const seed = randomBytes(32);
  const keypair = nacl.sign.keyPair.fromSeed(seed);
  return {
    private_key: toBase64Url(seed),
    public_key: toBase64Url(keypair.publicKey)
  };
}

export function ed25519PublicFromPrivate(signingPrivateKeyB64: string): string {
  const seed = asFixedBytes(fromBase64Url(signingPrivateKeyB64), 32, "Ed25519 private key");
  return toBase64Url(nacl.sign.keyPair.fromSeed(seed).publicKey);
}

export function generateX25519Keypair(): { private_key: string; public_key: string } {
  const keypair = nacl.box.keyPair();
  return {
    private_key: toBase64Url(keypair.secretKey),
    public_key: toBase64Url(keypair.publicKey)
  };
}

export function x25519PublicFromPrivate(encryptionPrivateKeyB64: string): string {
  const privateKey = asFixedBytes(fromBase64Url(encryptionPrivateKeyB64), 32, "X25519 private key");
  const keypair = nacl.box.keyPair.fromSecretKey(privateKey);
  return toBase64Url(keypair.publicKey);
}

export function signBytes(data: Uint8Array, signingPrivateKeyB64: string): string {
  const seed = asFixedBytes(fromBase64Url(signingPrivateKeyB64), 32, "Ed25519 private key");
  const signingKeypair = nacl.sign.keyPair.fromSeed(seed);
  const signature = nacl.sign.detached(data, signingKeypair.secretKey);
  return toBase64Url(signature);
}

export function verifySignature(
  data: Uint8Array,
  signatureB64: string,
  signingPublicKeyB64: string
): boolean {
  const signature = fromBase64Url(signatureB64);
  const publicKey = fromBase64Url(signingPublicKeyB64);
  if (signature.length !== nacl.sign.signatureLength || publicKey.length !== nacl.sign.publicKeyLength) {
    return false;
  }
  return nacl.sign.detached.verify(data, signature, publicKey);
}

export function envelopeAad(envelope: Envelope): Uint8Array {
  return canonicalJsonBytes({
    acp_version: envelope.acp_version,
    message_id: envelope.message_id,
    operation_id: envelope.operation_id,
    sender: envelope.sender,
    recipients: envelope.recipients
  });
}

function deriveWrapKey(sharedSecret: Uint8Array, recipient: string): Uint8Array {
  const derived = hkdfSync("sha256", sharedSecret, Buffer.alloc(0), `acp-v1-wrap:${recipient}`, 32);
  return new Uint8Array(derived);
}

function encryptAesGcm(
  key: Uint8Array,
  nonce: Uint8Array,
  plaintext: Uint8Array,
  aad: Uint8Array
): Uint8Array {
  const cipher = createCipheriv("aes-256-gcm", key, nonce);
  cipher.setAAD(aad);
  const ciphertext = Buffer.concat([cipher.update(plaintext), cipher.final()]);
  const authTag = cipher.getAuthTag();
  return Buffer.concat([ciphertext, authTag]);
}

function decryptAesGcm(
  key: Uint8Array,
  nonce: Uint8Array,
  ciphertextWithTag: Uint8Array,
  aad: Uint8Array
): Uint8Array {
  if (ciphertextWithTag.length < 16) {
    throw cryptoError("ciphertext is too short");
  }
  const ciphertext = ciphertextWithTag.subarray(0, ciphertextWithTag.length - 16);
  const tag = ciphertextWithTag.subarray(ciphertextWithTag.length - 16);
  const decipher = createDecipheriv("aes-256-gcm", key, nonce);
  decipher.setAAD(aad);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]);
}

export function encryptForRecipients(
  payload: JsonMap,
  envelope: Envelope,
  recipientEncryptionPublicKeys: Record<string, string>
): ProtectedPayload {
  const plaintext = canonicalJsonBytes(payload as JsonValue);
  const contentKey = randomBytes(32);
  const payloadNonce = randomBytes(12);
  const payloadCiphertext = encryptAesGcm(contentKey, payloadNonce, plaintext, envelopeAad(envelope));

  const ephemeral = nacl.box.keyPair();
  const wrappedContentKeys: WrappedContentKey[] = [];
  for (const [recipient, recipientPublicKeyB64] of Object.entries(recipientEncryptionPublicKeys)) {
    const recipientPublicKey = asFixedBytes(
      fromBase64Url(recipientPublicKeyB64),
      32,
      "X25519 public key"
    );
    const sharedSecret = nacl.scalarMult(ephemeral.secretKey, recipientPublicKey);
    const wrapKey = deriveWrapKey(sharedSecret, recipient);
    const wrapNonce = randomBytes(12);
    const wrappedContentKey = encryptAesGcm(
      wrapKey,
      wrapNonce,
      contentKey,
      new TextEncoder().encode(envelope.message_id)
    );
    wrappedContentKeys.push({
      recipient,
      ephemeral_public_key: toBase64Url(ephemeral.publicKey),
      nonce: toBase64Url(wrapNonce),
      ciphertext: toBase64Url(wrappedContentKey)
    });
  }

  return {
    nonce: toBase64Url(payloadNonce),
    ciphertext: toBase64Url(payloadCiphertext),
    wrapped_content_keys: wrappedContentKeys,
    payload_hash: sha256Hex(payloadCiphertext),
    signature_kid: "",
    signature: ""
  };
}

function messageSignatureInput(envelope: Envelope, protectedPayload: ProtectedPayload): Uint8Array {
  const signableProtected: JsonMap = {
    nonce: protectedPayload.nonce,
    ciphertext: protectedPayload.ciphertext,
    wrapped_content_keys: [...protectedPayload.wrapped_content_keys].sort((a, b) =>
      a.recipient.localeCompare(b.recipient)
    ) as unknown as JsonValue,
    payload_hash: protectedPayload.payload_hash,
    signature_kid: protectedPayload.signature_kid
  };
  return canonicalJsonBytes({
    envelope,
    protected: signableProtected
  } as unknown as JsonValue);
}

export function signProtectedPayload(
  envelope: Envelope,
  protectedPayload: ProtectedPayload,
  signingPrivateKeyB64: string,
  signatureKid: string
): ProtectedPayload {
  const signedPayload: ProtectedPayload = {
    ...protectedPayload,
    signature_kid: signatureKid
  };
  const signatureInput = messageSignatureInput(envelope, signedPayload);
  signedPayload.signature = signBytes(signatureInput, signingPrivateKeyB64);
  return signedPayload;
}

export function verifyProtectedPayloadSignature(
  envelope: Envelope,
  protectedPayload: ProtectedPayload,
  senderSigningPublicKeyB64: string
): boolean {
  if (!protectedPayload.signature.trim()) {
    return false;
  }
  const signatureInput = messageSignatureInput(envelope, protectedPayload);
  return verifySignature(signatureInput, protectedPayload.signature, senderSigningPublicKeyB64);
}

export function decryptForRecipient(
  envelope: Envelope,
  protectedPayload: ProtectedPayload,
  recipientId: string,
  recipientEncryptionPrivateKeyB64: string
): JsonMap {
  const wrapped = protectedPayload.wrapped_content_keys.find((item) => item.recipient === recipientId);
  if (!wrapped) {
    throw cryptoError(`No wrapped content key available for recipient ${recipientId}`);
  }
  const recipientPrivateKey = asFixedBytes(
    fromBase64Url(recipientEncryptionPrivateKeyB64),
    32,
    "X25519 private key"
  );
  const ephemeralPublicKey = asFixedBytes(
    fromBase64Url(wrapped.ephemeral_public_key),
    32,
    "X25519 public key"
  );
  const sharedSecret = nacl.scalarMult(recipientPrivateKey, ephemeralPublicKey);
  const wrapKey = deriveWrapKey(sharedSecret, recipientId);
  const wrappedNonce = asFixedBytes(fromBase64Url(wrapped.nonce), 12, "wrapped nonce");
  const wrappedCiphertext = fromBase64Url(wrapped.ciphertext);
  const contentKey = decryptAesGcm(
    wrapKey,
    wrappedNonce,
    wrappedCiphertext,
    new TextEncoder().encode(envelope.message_id)
  );
  const payloadNonce = asFixedBytes(fromBase64Url(protectedPayload.nonce), 12, "payload nonce");
  const payloadCiphertext = fromBase64Url(protectedPayload.ciphertext);
  const plaintext = decryptAesGcm(contentKey, payloadNonce, payloadCiphertext, envelopeAad(envelope));
  return parseJsonMap(new TextDecoder().decode(plaintext));
}
